### DISCUSSION TITLE
Typescript: How to properly type setValue

### DISCUSSION THREAD

**Discussion by azban** [react-hook-form/react-hook-form#7246] [General]
Hi, I was looking to create a wrapper function around setValue, but I can't figure out how to set up my generics to allow for setting a primitive value.

```ts
type FormInputs = {
  firstName: string;
  lastName: string;
}

const customSetValue = <T,>(
  inputName: Path<T>, 
  setValue: UseFormSetValue<T>,
) => {
  setValue(inputName, "string");
}
```

`Argument of type 'string' is not assignable to parameter of type 'UnpackNestedValue<PathValue<T, Path<T>>>'.ts(2345)`

Here's a super simple example of what I'm attempting to do: https://codesandbox.io/s/flamboyant-frog-dv8v1?file=/src/App.tsx

Is there a way to specify via generics that all values are string? Or for it to look up the value by the path?

**Comment by Joe312341** [comment]
Hey, there is an issue with path being able to properly use generics right now. I believe the RHF team is working on it, see https://github.com/react-hook-form/react-hook-form/pull/7215

**Comment by tommiCZE** [comment]
Hi i have same problem. Do u have some info about resolve this issue?

**Reply by azban** [reply]
nope, i had to use ts-ignore for now. was hoping #7215 would get merged, but looks like that was closed.

**Reply by Moshyfawn** [reply]
Take a look at the #7433 RHF V8 version changes. There will be a couple of Path changes that might be something that'll help

**Reply by tommiCZE** [reply]
Is know release date for RHF V8?

**Reply by Moshyfawn** [reply]
> Is know release date for RHF V8?

@bluebill1049

**Reply by bluebill1049** [reply]
No ETA at the moment.

**Comment by os-freddyshim** [comment]
What is the status on this issue? I came across the same problem and error messages and was wondering if a fix has been deployed yet.

**Comment by Moshyfawn** [comment]
The v8 is still in progress. You can already try the [v8.0.0-alpha.4](https://github.com/react-hook-form/react-hook-form/releases/tag/v8.0.0-alpha.4). No ETA for the stable release, tho

For now, the only thing I was able to hack together is this:
```ts
import type {
  UseFormSetValue,
  FieldValues,
  FieldPath,
  FieldPathValue
} from "react-hook-form";
import { useEffect } from "react";

type InputProps<
  TFieldValues extends FieldValues,
  TFieldName extends FieldPath<TFieldValues>
> = {
  setValue: UseFormSetValue<TFieldValues>
  name: TFieldName
};

const Input = <
  TFieldValues extends FieldValues,
  TFieldName extends FieldPath<TFieldValues>
>({
  setValue,
  name
}: InputProps<TFieldValues, TFieldName>) => {
  setValue(name, "string" as FieldPathValue<TFieldValues, TFieldName>)
};
```

Here's [a basic codesandbox](https://codesandbox.io/s/keen-matsumoto-99ccvx?file=%2Fsrc%2FApp.tsx)

**Reply by os-freddyshim** [reply]
I was able to get rid of Typescript errors using your approach after changing the casting of the string value (ie. I had to change FieldPathValue<TFieldValues, TFieldName> to UnpackNestedValue<FieldPathValue<TFieldValues, TFieldName>>).

```js
import {
  FieldValues,
  UseFormSetValue,
  FieldPath,
  FieldPathValue,
  UnpackNestedValue,
} from 'react-hook-form';

interface InputProps<T extends FieldValues> {
  name: FieldPath<T>;
  setValue: UseFormSetValue<T>;
}

const Input = <T extends FieldValues>({ name, setValue }: InputProps<T>) => {  
  setValue(
    name,
    'string' as UnpackNestedValue<FieldPathValue<T, FieldPath<T>>>
  );
};

export default Input;
```

**Reply by fresonn** [reply]
'UnpackNestedValue' is **deprecated**.

**Comment by fresonn** [comment]
setValue(name, value as PathValue<T, Path<T>>)

**Reply by iliapnmrv** [reply]
PathValue<T, Path> returns error `Generic type 'Path' requires 1 type argument(s)`
But PathValue<T, Path<T>> works fine, thank you!
